File: script/benchmark/benchmark.rb
#!/usr/bin/env ruby

# 如果测试普通Ruby，使用 ：ruby benchmark_ruby.rb
# 如果想测试jit，应该使用 sdk install ruby truffleruby 之后再ruby benchmark_ruby.rb

require 'benchmark'

def int_arith
  sum = 0
  1.upto(100_000_000) { |i| sum += i }
  sum
end

def float_arith
  sum = 0.0
  1.upto(10_000_000) { |i| sum += Math.sin(i) * Math.cos(i) }
  sum
end

def table_ops
  arr = Array.new(1_000_000) { |i| i }
  arr.sum
end

def recurse(n)
  return 0 if n == 0
  1 + recurse(n - 1)
end

Benchmark.bm(30) do |x|
  x.report("Integer Arithmetic")     { puts "Result: #{int_arith}" }
  x.report("Float Arithmetic")       { puts "Result: #{float_arith}" }
  x.report("Table Operations")       { puts "Result: #{table_ops}" }
  x.report("Recursion (Depth 5000)") { puts "Result: #{recurse(5000)}" }
end

File: script/benchmark/benchmark.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import math
import time

def benchmark(name, func):
    start = time.perf_counter()
    result = func()
    duration = time.perf_counter() - start
    print(f"[{name:<24}] {duration:.3f} seconds  Result: {result}")

# 整数加法
def int_arith():
    sum_ = 0
    for i in range(1, 100_000_001):
        sum_ += i
    return sum_

# 浮点运算
def float_arith():
    sum_ = 0.0
    for i in range(1, 10_000_001):
        sum_ += math.sin(i) * math.cos(i)
    return sum_

# 列表操作
def table_ops():
    arr = [i for i in range(1_000_000)]
    return sum(arr)

# 递归测试
def recurse(n):
    if n == 0:
        return 0
    return 1 + recurse(n - 1)

if __name__ == "__main__":
    benchmark("Integer Arithmetic", int_arith)
    benchmark("Float Arithmetic", float_arith)
    benchmark("Table Operations", table_ops)
    benchmark("Recursion (Depth 5000)", lambda: recurse(5000))
